### PR TITLE
GH-2: Add Streaming Support to SFTP Source

### LINKED ISSUE #2: Add streaming support for SFTP source
As a user, I'd like to use the streaming support for SFTP source, so I can stream all the file fragments instead of persisting it in the local disk and then transferring. 

**Acceptance:**
- sftp-source is retrofitted to also support streaming capability available in SI 4.3.5 
- test coverage included

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves #2 

Also, flatten properties to avoid missing descriptions in READMEs.

Requires https://github.com/spring-cloud-stream-app-starters/core/pull/15

### CODE REVIEW

File: spring-cloud-starter-stream-sink-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/sink/SftpSinkConfiguration.java [hunk 1]
```diff
@@ -58,4 +58,25 @@ public IntegrationFlow ftpInboundFlow(SftpSinkProperties properties, SessionFact
...
+	@ConditionalOnMissingBean
+	public SessionFactory<LsEntry> sftpSessionFactory(SftpSinkProperties properties) {
+		DefaultSftpSessionFactory sftpSessionFactory = new DefaultSftpSessionFactory();
+		sftpSessionFactory.setHost(properties.getFactory().getHost());
+		sftpSessionFactory.setPort(properties.getFactory().getPort());
+		sftpSessionFactory.setUser(properties.getFactory().getUsername());
+		sftpSessionFactory.setPassword(properties.getFactory().getPassword());
+		sftpSessionFactory.setAllowUnknownKeys(properties.getFactory().isAllowUnknownKeys());
+		if (properties.getFactory().getKnownHostsExpression() != null) {
+			sftpSessionFactory.setKnownHosts("#{" + properties.getFactory().getKnownHostsExpression() + "}");
```
Comment (artembilan) [inline spring-cloud-starter-stream-sink-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/sink/SftpSinkConfiguration.java]: SpEL?
```
/**
 * Specifies the filename that will be used for a host key repository.
 * The file has the same format as OpenSSH's known_hosts file.
 * <p>
 * <b>Required if {@link #setAllowUnknownKeys(boolean) allowUnknownKeys} is
 * false (default).</b>
 * @param knownHosts The known hosts.
 * @see JSch#setKnownHosts(String)
 */
public void setKnownHosts(String knownHosts) {
	this.knownHosts = knownHosts;
}
```

I don't see how it is resolved.
Thanks

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-cloud-starter-stream-sink-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/sink/SftpSinkConfiguration.java (hunk 1 above)
Comment (garyrussell) [reply spring-cloud-starter-stream-sink-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/sink/SftpSinkConfiguration.java]: I didn't look at the code, I just moved it - but you're right - no idea which brain cell generated that, but it must have died long ago - it looks like a bad port from XD.

Comment (artembilan) [reply spring-cloud-starter-stream-sink-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/sink/SftpSinkConfiguration.java]: Well, in XD that looks like:
```
<beans:property name="knownHosts" value="#{${knownHostsExpression:null}}" />
```
Which is really bean initialization SpEL, not a target property value like in this case.
There it has a description like:
```
@ModuleOption("a SpEL expresssion location of known hosts file; required if 'allowUnknownKeys' is false; "
			+ "examples: systemProperties[\"user.home\"]+\"/.ssh/known_hosts\", \"/foo/bar/known_hosts\"")
	public void setKnownHostsExpression(String knownHostsExpression) {
```
Which really makes sense. We can come up with something similar in Java config via `@Value`

File: spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/SftpSourceConfiguration.java [hunk 2]
```diff
@@ -48,50 +66,129 @@
...
-		});
+		};
+	}
 
-		return FileUtils.enhanceFlowForReadingMode(flowBuilder, fileConsumerProperties)
-				.channel(this.source.output())
-				.get();
+	private Consumer<SourcePollingChannelAdapterSpec> consumerSpecWithDelete(final SftpSourceProperties properties) {
+		final PollerMetadata poller = new PollerMetadata();
+		BeanUtils.copyProperties(this.defaultPoller, poller, "transactionSynchronizationFactory");
```
Comment (artembilan) [inline spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/SftpSourceConfiguration.java]: ???
Why do we need to copy that default one?
We are in the microservice here, so we have only singe `PollerMetadata` and just for this single channel adapter in the whole application. Therefore I think it is still safe just to enhance an existing `defaultPoller` without further `Consumer<SourcePollingChannelAdapterSpec>` creation. We still can rely on the `PollerMetadata.getDefaultPollerMetadata(this.beanFactory)` in the target `SourcePollingChannelAdapterFactoryBean`.

Comment (garyrussell) [reply spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/SftpSourceConfiguration.java]: I don't think we can make that assumption people might `@Import` this configuration.

**Comment by garyrussell** [comment]
I need to rebase and polish this after merging your other PR.

**Comment by garyrussell** [comment]
Rebased; fixed conflicts; added doc polish.

**Review by artembilan — APPROVED** [review]
LGTM
Merging...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-cloud-stream-app-starters/sftp/commit/9642029e5e36e0f4799689433399b24d5ef0c91d

**Pull request closed without merging**